xla/stream_executor/command_buffer.h

```diff
@@ -172,141 +105,126 @@ class CommandBuffer {
   // Command buffer API
   //===--------------------------------------------------------------------===//
 
-  // Adds an execution barrier to a given execution scope: all commands added
-  // before a barrier in a the execution scope will complete before any of the
-  // commands added after a barrier in the same execution scope.
-  virtual absl::Status Barrier(ExecutionScopeId execution_scope_id) = 0;
+  // Adds an execution barrier that depends on the commands in dep_idxes.
+  virtual absl::Status EmptyOp(Index cmd_idx, const CmdIndexSet& dep_idxes) = 0;
 
-  // Adds an execution barrier that synchronizes commands across multiple
-  // execution scopes. See example #2 in execution scope id documentation.
-  virtual absl::Status Barrier(
-      absl::Span<const ExecutionScopeId> execution_scope_ids) = 0;
+  // Adds an execution barrier that depends on all tail nodes (nodes that no
+  // other nodes depend on them) in current graph.
+  virtual absl::Status Barrier() = 0;
```

Comment:
One meta-comment for all APIs in this file:

We should not have two options, with and without id, what essentially each API call is doing is creating a new command (might be implemented as multiple internal nodes) and returns back an Id that allows you to depend on that command (for side effects ordering purpose).

So all command buffer APIs must have this signature:

absl::StatusOr<Id> CommandName(const Dependencies& deps, ....);